src/shaders/pbr.frag

```diff
@@ -0,0 +1,292 @@
+  vec3 F0 = mix(vec3(0.04), baseColor, metallic);
+  vec3 Fresnel = fresnelSchlick(F0, view, half);
+
+  // Diffuse BRDF
+  // NOTE: energy conservation requires
+  // diffuse + specular <= 1.0, where specular = Fresnel
+  // Also: result does not need to be scaled by 1/PI
+  // See details:
+  // https://seblagarde.wordpress.com/2012/01/08/pi-or-not-to-pi-in-game-lighting-equation/
+  vec3 diffuse = mix(vec3(1.0) - Fresnel, vec3(0.0), metallic) * baseColor;
+
+  // Specular BDDF
+  float temp =
+      max(4.0 * max(dot(normal, light), 0.0) * max(dot(normal, view), 0.0),
+          Epsilon);
+  vec3 specular = Fresnel *
+                  specularGeometricAttenuation(normal, light, view, roughness) *
+                  normalDistribution(normal, half, roughness) / temp;
+
+  return (diffuse + specular) * lightRadiance * max(dot(normal, light), 0.0);
+}
+
+void main() {
+  vec3 emissiveColor = Material.emissiveColor;
+#if defined(EMISSIVE_TEXTURE)
+  emissiveColor *= texture(emissiveTexture, texCoord).rgb;
+#endif
+  fragmentColor = vec4(emissiveColor, 0.0);
+
+#if (LIGHT_COUNT > 0)
+  vec4 baseColor = Material.baseColor;
+#if defined(BASECOLOR_TEXTURE)
+  baseColor *= texture(baseColorTexture, texCoord);
+#endif
+
+  float roughness = Material.roughness;
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+  roughness *= texture(packedTexture, texCoord).g;
+#elif defined(ROUGHNESS_TEXTURE)
+  roughness *= texture(roughnessTexture, texCoord).r;
+#endif
+
+  float metallic = Material.metallic;
+#if defined(NONE_ROUGHNESS_METALLIC_TEXTURE) || defined(OCCLUSION_ROUGHNESS_METALLIC_TEXTURE)
+  metallic *= texture(packedTexture, texCoord).b;
+#elif defined(METALLIC_TEXTURE)
+  metallic *= texture(metallicTexture, texCoord).r;
+#endif
+
+#if defined(NORMAL_TEXTURE)
+  // normal is now in the camera space
+  vec3 n = getNormalFromNormalMap();
+#else
+  vec3 n = normal;
+#endif
+
+  // view dirction: a vector from current position to camera
+  // in camera space, camera is at the origin
+  vec3 view = normalize(-position);
+
+  vec3 finalColor = vec3(0.0);
+  highp float finalAlpha = 0.0;
+  // compute contribution of each light using the microfacet model
+  // the following part of the code is inspired by the Phong.frag in Magnum
+  // library (https://magnum.graphics/)
+  for (int iLight = 0; iLight < LIGHT_COUNT; ++iLight) {
+    // Attenuation. Directional lights have the .w component set to 0, use
+    // that to make the distance zero -- which will then ensure the
+    // attenuation is always 1.0
+    highp float dist =
+        length(LightDirections[iLight].xyz) * LightDirections[iLight].w;
+    // If range is 0 for whatever reason, clamp it to a small value to
+    // avoid a NaN when dist is 0 as well (which is the case for
+    // directional lights).
+    highp float attenuation =
+        clamp(1.0 - pow(dist / max(LightRanges[iLight], 0.0001), 4.0), 0.0,
+              1.0) /
+        (1.0 + dist);
```

Comment:
Nit: @mosra  , I saw you did this divide by `(1.0 + dist)` in the phong shader, too. `1 / (1.0 + dist)^2` is not quite the same as `1 / dist^2`. In particular, it behaves very differently for small distances. Isn't this deviating from the standard falloff calculation?

Reply:
Yeah I spent some time investigating this already ([check out the docs](https://doc.magnum.graphics/magnum/classMagnum_1_1Trade_1_1LightData.html#Trade-LightData-attenuation)) -- If you don't add the 1, it'll explode to infinity at `dist` close to 0, which I don't think is desired :) 

For some reason the glTF spec [doesn't include that in the equation](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_lights_punctual#range-property), but the UE4 calculation (the glTF equation was based on) [has it](http://blog.selfshadow.com/publications/s2013-shading-course/karis/s2013_pbs_epic_notes_v2.pdf):

![image](https://user-images.githubusercontent.com/344828/95613919-f565db80-0a65-11eb-8f43-7efbe563cbd4.png)

Hmm ... but it should be `(1.0 + dist^2)` not `(1.0 + dist)^2`. I have to check why I did that.

Reply:
@eundersander : You concern sounds reasonable.

About the attenuation:
Here, I followed what was used in the Phong shader in Magnum. When I coded it up, I understand the `1.0` is to avoid division by 0.
I did not dig into it further. But if the paper says it should be `1 + dist^2`, I will change it.

Reply:
I just updated the code using the `1+dist^2` now. Take a look.

Reply:
Okay, then I'll do the same change in Phong. Will get back to you if I discover any issues with that.